pkg/controller/clusterdeployment/clusterdeployment_controller.go

```diff
@@ -1309,3 +1332,79 @@ func initializeAnnotations(cd *hivev1.ClusterDeployment) {
 		cd.Annotations = map[string]string{}
 	}
 }
+
+// managePullSecrets() takes pull secretes in to account and returns the merged pull request
+func (r *ReconcileClusterDeployment) managePullSecrets(cd *hivev1.ClusterDeployment, cdLog log.FieldLogger) (string, error) {
+
+	// For code readability let's call the pull secret in cluster deployment config as local pull secret
+	localPullSecret, err := controllerutils.LoadSecretData(r.Client, cd.Spec.PullSecret.Name, cd.Namespace, corev1.DockerConfigJsonKey)
+	if err != nil {
+		cdLog.Warn("unable to load pull secret from secret")
+	}
+
+	// Check if global pull secret from env as it comes from hive config
+	globalPullSecret := os.Getenv("GLOBAL_PULL_SECRET")
+	if len(globalPullSecret) == 0 {
+		cdLog.Warn("global pull secret not available")
+		if localPullSecret == "" {
+			errMsg := "Pull request or global pull request not available"
+			cdLog.Error(errMsg)
+			return "", fmt.Errorf("%s", errMsg)
+		}
+	}
+
+	// If global pull secret is present then merge the pullSecret and globalPullSecret
+	if globalPullSecret != "" && localPullSecret != "" {
+		// Merge local pullSecret and globalPullSecret. If both pull secrets have same registry name
+		// then the merged pull secret will have registry secret from local pull secret
+		pullSecret, err := controllerutils.MergeJsons(globalPullSecret, localPullSecret)
+		if err != nil {
+			errMsg := "unable to merge global pull secret with local pull secret"
+			cdLog.WithError(err).Error(errMsg)
+			return "", fmt.Errorf("%s : %v", errMsg, err)
+		}
+		return pullSecret, nil
+	} else if globalPullSecret != "" {
+		return globalPullSecret, nil
+	} else {
+		return localPullSecret, nil
+	}
+}
+
+// updatePullSecretObj updates pull secret object in cluster deployment status if required
+func (r *ReconcileClusterDeployment) updatePullSecretObj(pullSecret string, cd *hivev1.ClusterDeployment, cdLog log.FieldLogger) error {
+
+	// Check if pull secret object already exists
+	existingPullSecretObj := r.getExistingPullSecret(cd)
+	pullSecretObj := generatePullSecretObj(pullSecret, cd)
+
+	var pullSecretHash, existingPullSecretHash string
+
+	if existingPullSecretObj != nil {
+		existingPullSecretHash = controllerutils.CalculateHashOfSecret(existingPullSecretObj)
+		pullSecretHash = controllerutils.CalculateHashOfSecret(pullSecretObj)
+	}
+	if pullSecretHash != existingPullSecretHash {
+		if err := controllerutil.SetControllerReference(cd, pullSecretObj, r.scheme); err != nil {
+			cdLog.WithError(err).Error("error setting controller reference on pullSecretObj")
+			return err
+		}
+		err := r.Update(context.TODO(), pullSecretObj)
```

Comment:
Do we need to `Create` in the case where there is no existing pull secret?

Reply:
I am assumed update will create if it does not exist. But need to check that.

Reply:
fixed it.